skimage/morphology/_util.py

```diff
@@ -51,35 +54,208 @@ def _validate_connectivity(image_dim, connectivity, offset):
+
+    # If any dimension in image_shape is smaller than selem.shape
+    # duplicates might occur, remove them
+    if any(x < y for x, y in zip(image_shape, selem.shape)):
+        # np.unique reorders, which we don't want
+        _, indices = np.unique(raveled_offsets, return_index=True)
+        raveled_offsets = raveled_offsets[np.sort(indices)]
+
+    # Remove "offset to center"
+    raveled_offsets = raveled_offsets[1:]
+
+    return raveled_offsets
+
+
+def _resolve_neighborhood(selem, connectivity, ndim):
+    """Validate or create structuring element.
+
+    Depending on the values of `connectivity` and `selem` this function
+    either creates a new structuring element (`selem` is None) using
+    `connectivity` or validates the given structuring element (`selem` is not
+    None).
+
+    Parameters
+    ----------
+    selem : ndarray
+        A structuring element used to determine the neighborhood of each
+        evaluated pixel (``True`` denotes a connected pixel). It must be a
+        boolean array and have the same number of dimensions as `image`. If
+        neither `selem` nor `connectivity` are given, all adjacent pixels are
+        considered as part of the neighborhood.
+    connectivity : int
+        A number used to determine the neighborhood of each evaluated pixel.
+        Adjacent pixels whose squared distance from the center is less than or
+        equal to `connectivity` are considered neighbors. Ignored if
+        `selem` is not None.
+    ndim : int
+        Number of dimensions `selem` ought to have.
+
+    Returns
+    -------
+    selem : ndarray
+        Validated or new structuring element specifying the neighborhood.
+
+    Examples
+    --------
+    >>> _resolve_neighborhood(None, 1, 2)
+    array([[False,  True, False],
+           [ True,  True,  True],
+           [False,  True, False]])
+    >>> _resolve_neighborhood(None, None, 3).shape
+    (3, 3, 3)
     """
-    structure = structure.copy()  # Don't modify original input
-    structure[tuple(center)] = 0  # Ignore the center; it's not a neighbor
-    connection_indices = np.transpose(np.nonzero(structure))
-    offsets = (np.ravel_multi_index(connection_indices.T, image_shape,
-                                    order=order) -
-               np.ravel_multi_index(center, image_shape, order=order))
-    squared_distances = np.sum((connection_indices - center) ** 2, axis=1)
-    return offsets[np.argsort(squared_distances)]
+    if selem is None:
+        if connectivity is None:
+            connectivity = ndim
+        selem = ndi.generate_binary_structure(ndim, connectivity)
+    else:
+        # Validate custom structured element
+        selem = np.asarray(selem, dtype=np.bool)
+        # Must specify neighbors for all dimensions
+        if selem.ndim != ndim:
+            raise ValueError(
+                "structuring element and image must have the same number of "
+                "dimensions"
```

Comment:
```suggestion
                "Structuring element and image must have the same number of "
                "dimensions."
```

Reply:
This deliberately emulates what error messages from Python and NumPy itself look like: a concise statement about "what's wrong" without punctuation. So I'd prefer to keep it this way. Same for some of you other comments.